src/9FC03F10.c

```diff
@@ -1,15 +1,140 @@
 #include "include_asm.h"
+#include "bcp.h"
+#include "ultratypes.h"
 
-INCLUDE_ASM("asm/non_matchings/9FC03F10", delay);
+void delay(s32 arg0) {
+    s32 i;
 
-INCLUDE_ASM("asm/non_matchings/9FC03F10", initialize_virage_controller);
+    for(i = 0; i < arg0; i++);
+}
 
-INCLUDE_ASM("asm/non_matchings/9FC03F10", initialize_virage_controllers);
+void initialize_virage_controller(void* controller) {
+    s32 temp = (u32)controller & 0xFFFF0000;
 
-INCLUDE_ASM("asm/non_matchings/9FC03F10", write_virage_data);
+    IO_WRITE(temp + 0x8000, 0x8A);
+    IO_WRITE(temp + 0x8004, 0x13);
+    IO_WRITE(temp + 0x8008, 0x80);
+    IO_WRITE(temp + 0x800C, 0x92);
+    IO_WRITE(temp + 0x8010, 0x18);
+    IO_WRITE(temp + 0x8014, 5);
+}
 
-INCLUDE_ASM("asm/non_matchings/9FC03F10", func_9FC04220);
+void initialize_virage_controllers(void) {
+    IO_WRITE(VIRAGE2_STATUS_REG, 0);
+    initialize_virage_controller(VIRAGE0_STATUS_REG);
+    initialize_virage_controller(VIRAGE1_STATUS_REG);
+    initialize_virage_controller(VIRAGE2_STATUS_REG);
+}
 
-INCLUDE_ASM("asm/non_matchings/9FC03F10", func_9FC0425C);
+s32 write_virage_data(void* controller, s32 *data, s32 size) {
+    s32 temp_a0;
+    s32 temp_a1;
+    s32 temp_lo;
+    s32 temp_s1;
+    u32 temp_s4;
+    s32 i;
+    u32 temp;
+    u32 temp2;
 
-INCLUDE_ASM("asm/non_matchings/9FC03F10", func_9FC04304);
+    temp_s4 = (u32)controller & 0xFFFF0000;
+    temp_s1 = func_9FC04220();
+    IO_WRITE(controller, 0);
+    IO_WRITE(MI_1C_REG, (1000 / temp_s1) + 1);
+    delay(640000 / temp_s1);
+    temp = IO_READ(controller);
+    if (temp & 1) {
+        temp_lo = 20000 / temp_s1;
+        delay(temp_lo);
+        delay(temp_lo);
+        temp = IO_READ(controller);
+        if (temp & 1) {
+	        return -1;
+        }
+    }
+
+    temp = IO_READ(controller);
+    if ((temp & 0x40000000) == 0) {
```

Comment:
```suggestion
    if (!(temp & 0x40000000)) {
```
Very minor but I think these kinds of checks are a bit nicer like this